src/dataflow/DataflowAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,379 @@
+        TypeElement typeElt = types.boxedClass(type.getUnderlyingType());
+        AnnotationMirror am = DataflowUtils.createDataflowAnnotation(typeElt.asType().toString(),
+                this.processingEnv);
+        AnnotatedDeclaredType dt = fromElement(typeElt);
+        dt.addAnnotation(am);
+        return dt;
+    }
+
+    /**
+     * This method handles unboxing for reference type.
+     * For statements, int i = new Integer(3);
+     * The annotation for i should be @DataFlow(typeNames = {"int"}).
+     */
+    @Override
+    public AnnotatedPrimitiveType getUnboxedType(AnnotatedDeclaredType type)
+            throws IllegalArgumentException {
+        PrimitiveType primitiveType = types.unboxedType(type.getUnderlyingType());
+        AnnotationMirror am = DataflowUtils.createDataflowAnnotation(primitiveType.toString(),
+                this.processingEnv);
+        AnnotatedPrimitiveType pt = (AnnotatedPrimitiveType) AnnotatedTypeMirror.createType(
+                primitiveType, this, false);
+        pt.addAnnotation(am);
+        return pt;
+    }
+
+    private final class DataFlowQualifierHierarchy extends GraphQualifierHierarchy {
+
+        public DataFlowQualifierHierarchy(MultiGraphFactory f, AnnotationMirror bottom) {
+            super(f, bottom);
+        }
+
+        /**
+         * This method checks whether rhs is subtype of lhs. rhs and lhs are
+         * both Dataflow types with typeNameRoots argument.
+         * 
+         * @param rhs
+         * @param lhs
+         * @return true is rhs is subtype of lhs, otherwise return false.
+         */
+        private boolean isSubtypeWithRoots(AnnotationMirror rhs, AnnotationMirror lhs) {
+
+            Set<String> rTypeNamesSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNames(rhs)));
+            Set<String> lTypeNamesSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNames(lhs)));
+            Set<String> rRootsSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNameRoots(rhs)));
+            Set<String> lRootsSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNameRoots(lhs)));
+            Set<String> combinedTypeNames = new HashSet<String>();
+            combinedTypeNames.addAll(rTypeNamesSet);
+            combinedTypeNames.addAll(lTypeNamesSet);
+            Set<String> combinedRoots = new HashSet<String>();
+            combinedRoots.addAll(rRootsSet);
+            combinedRoots.addAll(lRootsSet);
+
+            AnnotationMirror combinedAnno = DataflowUtils.createDataflowAnnotationWithRoots(
+                    combinedTypeNames, combinedRoots, processingEnv);
+            AnnotationMirror refinedCombinedAnno = refineDataflow(combinedAnno);
+            AnnotationMirror refinedLhs = refineDataflow(lhs);
+
+            if (AnnotationUtils.areSame(refinedCombinedAnno, refinedLhs)) {
```

Comment:
Hi @Jianchu , I have some questions about the type hierarchy in dataflow system.

Refer to your thesis, it says `typeNameRoots` is "all possible upper bounds of Java run-time types a method may return" and `typeName` is "all possible Java run-time types a method may return". And it declared `@Dataflow(typeNameRoots="Object") is the top qualifier.

My questions are:

1. Given a Dataflow annotation, should it always follows: for every value in `typeName` field, there is an upper bound in `typeNameRoots` field will be present. E.g. `@Dataflow(roots={Number, List}, names={Integer, ArrayList})` is valid. But `@Dataflow(roots={Number}, names={Object})` is not valid, because `roots` should be "all possible upper bound of run-time types" of the annotated method return type, but this type might be `Object`(as shown in `names` field), which is not upper bounded by any values in `roots` field.

2. As your thesis claimed, `@Dataflow(roots=Object)` is the top qualifier. However, according to your `subtype()` implementation here, type `T1` is a super type of `T2`, if and only if `T1`'s `roots` and `names` fields are BOTH superset of `T2`'s `roots` and `names`. Then there is a contradiction between your thesis and your implementation, for example: `@Dataflow(roots=Object)` is not a super type of `@Dataflow(roots="", names="String")`, as top qualifier's `names` field is empty, which is not a superset of second qualifier's `names` field.